### PR TITLE
AD-ify thermal expansion function methods

### PR DESCRIPTION
**Pull request by tophmatthews**
This creates AD verisions of `ADComputeMeanThermalExpansionFunctionEigenstrain`, `ADComputeInstantaneousThermalExpansionFunctionEigenstrain`, and `ADComputeMeanThermalExpansionEigenstrainBase`. The first two do not completely utilize the AD system since there is no current method to extract dual number information from a function (I think). They are really in place to test the base class, which is needed elsewhere. AD classes that derive from `ADComputeMeanThermalExpansionEigenstrainBase` will be able to take full advantage of the AD system, since the thermal expansion calculation can be entirely defined up in the inherited class.

Although the two "function" classes don't utilize AD, they may serve as good examples of where functions that take advantage of AD classes can be helpful.

### CODE REVIEW

**Comment by tophmatthews** [comment]
@lindsayad @dschwen This may be a good example of where dual number information from functions could help. This PR really was motivated by the need for the base class, but the functions are used to test the base class in the non-AD versions. The AD verisions of the function classes currently don't really use AD due to the function value restriction, but they still test out the base class.

Although this may help define the need for AD functions, I still would like the base class in before too long as it is holding up a PR in a different app.

**Comment by lindsayad** [comment]
AD functions are on my radar; hopefully get to it not too long after https://github.com/libMesh/libmesh/pull/2396

**Comment by tophmatthews** [comment]
Good to hear. Anyway we can get this in so I can start using the base class, and then circle back when functions are ready @lindsayad ?

**Comment by lindsayad** [comment]
Sure

> Good to hear. Anyway we can get this in so I can start using the base class, and then circle back when functions are ready @lindsayad ?

I would think so. I think a mechanics person should review this though

**Comment by tophmatthews** [comment]
@dschwen can you take a look at this since you did most of the underlying pieces?

File: modules/tensor_mechanics/src/materials/ADComputeMeanThermalExpansionEigenstrainBase.C
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ADComputeMeanThermalExpansionEigenstrainBase.C]: Is this used?

File: modules/tensor_mechanics/src/materials/ADComputeMeanThermalExpansionFunctionEigenstrain.C
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ADComputeMeanThermalExpansionFunctionEigenstrain.C]: We should warn somewhere that this drops all derivative info (you might do that already and I just missed it)

**Review by dschwen — COMMENTED** [review]
This looks fine, except  that I don't really get the point of this, besides being ready for upcoming ADFunctions. Is that it?

**Comment by tophmatthews** [comment]
@dschwen I needed the base class for another application. The function methods are used to test the non-AD pieces, so I duplicated them in AD to make the tests identical. Sure, the function pieces don't work yet, but they're good examples, while still providing the base class.

However, I am thinking this may be unnecessary, and I should be using `ADComputeThermalExpansionEigenstrainBase` instead...but I'm not positive...

**Comment by tophmatthews** [comment]
Let me circle back on this. This is still a fine addition, just may not be needed as quick as I thought.

**Comment by tophmatthews** [comment]
@dschwen I kept much of the previous stuff of this PR, and added `ComputeDilatationThermalExpansionEigenstrainBase`, which is a base class that uses the relationships for dilatation, and `ComputeDilatationThermalExpansionFunctionEigenstrain` which is used to test the base class. This is really what I needed to do for other applications, but I didn't want to throw away all the rest of the work. I made AD versions for the dilatation models as well. In additions, I added alerts to the document pages for all AD function methods here that calls out the fact that the jacobians will be imperfect until AD functions are in.

**Comment by tophmatthews** [comment]
This will close #14595

**Comment by tophmatthews** [comment]
Here, dilatation is just the (delta L / L0) equation that is often used to describe experimental data. Casting it into a mean thermal expansion coefficient, and then reversing it back to get a (delta L/L0) eigenstrain was overkill.

**Comment by tophmatthews** [comment]
Just a reminder @dschwen after your likely powder filled weekend...

**Review by dschwen — APPROVED** [review]

**Comment by tophmatthews** [comment]
This is failing the push to devel for documentation, but I have no idea why @dschwen @permcody ?
I think it is the `MooseDocs.extensions.civet (MainProcess): 'generate_test_reports' requires results to exist.` line, but how do I fix that?
```
======================================================
MooseDocs.Translator (MainProcess): Executing extension init() methods...
MooseDocs.extensions.civet (MainProcess): Collecting CIVET results...
MooseDocs.extensions.civet (MainProcess): Collecting CIVET results complete [0.18363308906555176 sec.]
MooseDocs.extensions.civet (MainProcess): 'generate_test_reports' requires results to exist.
MooseDocs.Translator (MainProcess): Executing extension init() methods complete [0.18497467041015625 sec.]
MooseDocs.extensions.appsyntax (MainProcess): Reading MOOSE application syntax...
MooseDocs.extensions.appsyntax (MainProcess): MOOSE application syntax complete [5.045456171035767 sec.]
MooseDocs.extensions.appsyntax (MainProcess): Building MOOSE class database...
MooseDocs.extensions.appsyntax (MainProcess): MOOSE class database complete [5.0982749462127686 sec]
MooseDocs.commands.check (MainProcess): A page for /Functions exists, but it is still listed as hidden.
MooseDocs.commands.check (MainProcess): A page for /GlobalParams exists, but it is still listed as hidden.
MooseDocs.commands.check (MainProcess): A page for /AuxKernels/BoundsAux exists, but it is still listed as hidden.
MooseDocs.commands.check (MainProcess): A MOOSE generated stub page for /AuxKernels/ElementIntegerAux exists, but no content was added. Add documentation content to /opt/civet/build_0/moose/framework/doc/content/source/auxkernels/ElementIntegerAux.md.
MooseDocs.commands.check (MainProcess): A MOOSE generated stub page for /Executioner/Adaptivity exists, but no content was added. Add documentation content to /opt/civet/build_0/moose/framework/doc/content/syntax/Executioner/Adaptivity/index.md.
MooseDocs.Executioner (MainProcess): Executing preExecute methods...
MooseDocs.Executioner (MainProcess): Finished preExecute methods [0.0760035514831543 sec.]
MooseDocs.Executioner (MainProcess): Translating using 24 threads...
MooseDocs.extensions.appsyntax (Process-17): The class description is missing for /Postprocessors/GreaterThanLessThanPostprocessor, it can be added using the 'addClassDescription' method from within the objects validParams function.
MooseDocs.Executioner (MainProcess): Translating complete [69.69279313087463 sec.]
MooseDocs.Executioner (MainProcess): Copying content...
MooseDocs.Executioner (MainProcess): Copying Finished [0.8411340713500977 sec.]
MooseDocs.Executioner (MainProcess): Executing postExecute methods...
MooseDocs.Executioner (MainProcess): Finished postExecute methods [0.00020813941955566406 sec.]
MooseDocs.Executioner (MainProcess): Total Time [70.612473487854 sec.]
CRITICAL:0 ERROR:1 WARNING:6
Makefile:32: recipe for target 'test_doc' failed
make: *** [test_doc] Error 1
```

**Comment by permcody** [comment]
I don't see the error message in this text. I believe those specific issues
are just warnings.

**Pull request merged**
